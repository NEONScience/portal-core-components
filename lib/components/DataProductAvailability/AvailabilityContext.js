"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getTestableItems = exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _cloneDeep = _interopRequireDefault(require("lodash/cloneDeep"));
var _NeonContext = _interopRequireDefault(require("../NeonContext/NeonContext"));
var _NeonEnvironment = _interopRequireDefault(require("../NeonEnvironment/NeonEnvironment"));
var _AvailabilityUtils = require("./AvailabilityUtils");
var _NeonSignInButtonState = _interopRequireDefault(require("../NeonSignInButton/NeonSignInButtonState"));
var _StateStorageService = _interopRequireDefault(require("../../service/StateStorageService"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const SORT_DIRECTIONS = {
  ASC: 'ASC',
  DESC: 'DESC'
};
const DEFAULT_STATE = {
  sites: [],
  tables: {},
  rows: {},
  rowTitles: {},
  rowLabels: [],
  breakouts: [],
  validBreakouts: ['states', 'domains', 'sites', 'tables'],
  sortDirection: SORT_DIRECTIONS.ASC,
  neonContextHydrated: false,
  // Whether NeonContext data has been one-time hydrated into state
  reference: {
    sites: {},
    states: {},
    domains: {}
  }
};

/**
   Function to copy necessary NeonContext data into local state when ready
*/
const hydrateNeonContextData = (state, neonContextData) => {
  const newState = _extends({}, state, {
    neonContextHydrated: true
  });
  newState.reference.sites = neonContextData.sites || {};
  newState.reference.states = neonContextData.states || {};
  newState.reference.domains = neonContextData.domains || {};
  newState.neonContextHydrated = true;
  return newState;
};

/**
   Extract Tables
*/
const extractTables = state => {
  const tables = {};
  state.sites.forEach(site => {
    const {
      siteCode
    } = site;
    site.tables.forEach(table => {
      const {
        name,
        description,
        waitInterval
      } = table;
      if (!tables[name]) {
        tables[name] = {
          name,
          description,
          waitInterval,
          sites: [siteCode]
        };
      } else if (!tables[name].sites.includes(siteCode)) {
        tables[name].sites.push(siteCode);
      }
    });
  });
  const digits = Math.floor(Math.log10(Object.keys(tables).length) + 1);
  Object.keys(tables).forEach((table, idx) => {
    tables[table].tableCode = "T".concat((idx + 1).toString().padStart(digits, '0'));
  });
  return tables;
};

/**
  Calculate Rows
  Take a state object and spit out a new state containing rows for use in the SVG grid
*/
const calculateRows = state => {
  if (!state.neonContextHydrated) {
    return state;
  }
  const newState = _extends({}, state, {
    rows: {},
    rowLabels: [],
    rowTitles: {}
  });
  const {
    sites,
    breakouts,
    reference,
    sortDirection,
    tables: knownTables
  } = state;
  if (!breakouts.length) {
    newState.rowLabels = ['ALL'];
    newState.rows.ALL = {};
  }
  // Generate unsorted row labels and row data respecting breakouts
  sites.forEach(site => {
    const {
      siteCode,
      tables
    } = site;
    if (!reference.sites[siteCode]) {
      return;
    }
    const {
      stateCode,
      domainCode
    } = reference.sites[siteCode];
    tables.forEach(table => {
      const {
        name: tableName,
        months
      } = table;
      if (!knownTables[tableName]) {
        return;
      }
      const {
        tableCode
      } = knownTables[tableName];
      // turn siteCode / stateCode / domainCode / tableCode into a label
      // then populate that label's row
      let title = !breakouts.length ? 'Summary of all sites and tables' : '';
      const label = !breakouts.length ? 'ALL' : breakouts.reduce((acc, cur) => {
        switch (cur) {
          case 'states':
            title = "".concat(title).concat(title.length ? ' - ' : '').concat(reference.states[stateCode].name);
            return "".concat(acc).concat(acc.length ? '-' : '').concat(stateCode);
          case 'domains':
            title = "".concat(title).concat(title.length ? ' - ' : '').concat(reference.domains[domainCode].name);
            return "".concat(acc).concat(acc.length ? '-' : '').concat(domainCode);
          case 'sites':
            title = "".concat(title).concat(title.length ? ' - ' : '').concat(reference.sites[siteCode].description);
            return "".concat(acc).concat(acc.length ? '-' : '').concat(siteCode);
          case 'tables':
            title = "".concat(title).concat(title.length ? ' - ' : '').concat(tableName);
            return "".concat(acc).concat(acc.length ? '-' : '').concat(tableCode);
          default:
            return acc;
        }
      }, '');
      if (!newState.rowLabels.includes(label)) {
        newState.rowLabels.push(label);
        newState.rows[label] = {};
      }
      newState.rowTitles[label] = title;
      Object.keys(months).forEach(month => {
        newState.rows[label][month] = [...(newState.rows[label][month] || []), months[month]];
      });
    });
    // Sort row labels
    const sortReturns = [sortDirection === SORT_DIRECTIONS.ASC ? 1 : -1, sortDirection === SORT_DIRECTIONS.ASC ? -1 : 1];
    newState.rowLabels.sort((a, b) => a < b ? sortReturns[0] : sortReturns[1]);
  });
  return newState;
};

/**
   Reducer
*/
const reducer = (state, action) => {
  const newState = _extends({}, state);
  switch (action.type) {
    case 'hydrateNeonContextData':
      if (!action.neonContextData) {
        return state;
      }
      return calculateRows(hydrateNeonContextData(newState, action.neonContextData));
    case 'setBreakouts':
      if (!action.breakouts) {
        return state;
      }
      if (!action.breakouts.every(b => state.validBreakouts.includes(b))) {
        return state;
      }
      newState.breakouts = [...action.breakouts];
      // If new breakouts contains the old leading breakout (the sort method) then keep it in front
      if (state.breakouts.length && newState.breakouts.includes(state.breakouts[0])) {
        newState.breakouts = newState.breakouts.filter(b => b !== state.breakouts[0]);
        newState.breakouts.unshift(state.breakouts[0]);
      }
      return calculateRows(newState);
    case 'setSortMethod':
      // Sort method is not a top-level value; rather the first breakout in the list.
      if (!state.validBreakouts.includes(action.method)) {
        return state;
      }
      newState.breakouts = state.breakouts.filter(b => b !== action.method);
      newState.breakouts.unshift(action.method);
      return calculateRows(newState);
    case 'setSortDirection':
      if (!Object.keys(SORT_DIRECTIONS).includes(action.direction)) {
        return state;
      }
      newState.sortDirection = action.direction;
      return calculateRows(newState);

    // Default
    default:
      return state;
  }
};

/**
   Context and Hook
*/
const Context = /*#__PURE__*/(0, _react.createContext)(DEFAULT_STATE);
const useAvailabilityState = () => {
  const hookResponse = (0, _react.useContext)(Context);
  if (hookResponse.length !== 2) {
    return [(0, _cloneDeep.default)(DEFAULT_STATE), () => {}];
  }
  return hookResponse;
};

/**
 * Defines a lookup of state key to a boolean
 * designating whether or not that instance of the context
 * should pull the state from the session storage and restore.
 * Keeping this lookup outside of the context provider function
 * as to not incur lifecycle interference by storing with useState.
 */
const restoreStateLookup = {};

/**
   Context Provider
*/
const Provider = props => {
  const {
    sites,
    dataAvailabilityUniqueId,
    children
  } = props;
  const [{
    data: neonContextData,
    isFinal: neonContextIsFinal,
    hasError: neonContextHasError
  }] = _NeonContext.default.useNeonContextState();
  const key = "availabilityContextState-".concat(dataAvailabilityUniqueId);
  if (typeof restoreStateLookup[key] === 'undefined') {
    restoreStateLookup[key] = true;
  }
  const shouldRestoreState = restoreStateLookup[key];
  const stateStorage = (0, _StateStorageService.default)(key);
  const savedState = stateStorage.readState();

  /**
     Initial State and Reducer Setup
  */
  let initialState = _extends({}, (0, _cloneDeep.default)(DEFAULT_STATE), {
    sites
  });
  initialState.tables = extractTables(initialState);
  if (neonContextIsFinal && !neonContextHasError && !savedState) {
    initialState = hydrateNeonContextData(initialState, neonContextData);
  }
  if (savedState && shouldRestoreState) {
    restoreStateLookup[key] = false;
    stateStorage.removeState();
    initialState = savedState;
  }
  const [state, dispatch] = (0, _react.useReducer)(reducer, calculateRows(initialState));

  // The current sign in process uses a separate domain. This function
  // persists the current state in storage when the button is clicked
  // so the state may be reloaded when the page is reloaded after sign
  // in.
  (0, _react.useEffect)(() => {
    const subscription = _NeonSignInButtonState.default.getObservable().subscribe({
      next: () => {
        if (!_NeonEnvironment.default.enableGlobalSignInState) return;
        restoreStateLookup[key] = false;
        stateStorage.saveState(state);
      }
    });
    return () => {
      subscription.unsubscribe();
    };
  }, [state, stateStorage, key]);

  /**
     Effect - Watch for changes to NeonContext data and push into local state
  */
  (0, _react.useEffect)(() => {
    if (!state.neonContextHydrated && neonContextIsFinal && !neonContextHasError) {
      dispatch({
        type: 'hydrateNeonContextData',
        neonContextData
      });
    }
  }, [neonContextData, neonContextIsFinal, neonContextHasError, state.neonContextHydrated, dispatch]);

  /**
     Render
  */
  return (
    /*#__PURE__*/
    // eslint-disable-next-line react/jsx-no-constructed-context-values
    _react.default.createElement(Context.Provider, {
      value: [state, dispatch]
    }, children)
  );
};
Provider.propTypes = {
  dataAvailabilityUniqueId: _propTypes.default.number,
  sites: _AvailabilityUtils.AvailabilityPropTypes.enhancedSites,
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.string])), _propTypes.default.node, _propTypes.default.string]).isRequired
};
Provider.defaultProps = {
  dataAvailabilityUniqueId: 0,
  sites: []
};

/**
   Export
*/
const AvailabilityContext = {
  Provider,
  useAvailabilityState,
  SORT_DIRECTIONS
};
var _default = exports.default = AvailabilityContext; // Additional items exported for unit testing
const getTestableItems = () => process.env.NODE_ENV !== 'test' ? {} : {
  DEFAULT_STATE,
  calculateRows,
  extractTables,
  hydrateNeonContextData,
  reducer
};
exports.getTestableItems = getTestableItems;