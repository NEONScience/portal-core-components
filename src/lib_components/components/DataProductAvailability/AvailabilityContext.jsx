import React, {
  createContext,
  useContext,
  useEffect,
  useReducer,
} from 'react';
import PropTypes from 'prop-types';

import cloneDeep from 'lodash/cloneDeep';

import NeonContext from '../NeonContext/NeonContext';

import { AvailabilityPropTypes } from './AvailabilityUtils';

const SORT_DIRECTIONS = { ASC: 'ASC', DESC: 'DESC' };
const DEFAULT_STATE = {
  sites: [],
  tables: {},
  rows: {},
  rowTitles: {},
  rowLabels: [],
  breakouts: [],
  validBreakouts: ['states', 'domains', 'sites', 'tables'],
  sortDirection: SORT_DIRECTIONS.ASC,
  neonContextLoaded: false,
  allSites: {},
  allStates: {},
  allDomains: {},
};

/**
   Extract Tables
*/
const extractTables = (state) => {
  const tables = {};
  state.sites.forEach((site) => {
    const { siteCode } = site;
    site.tables.forEach((table) => {
      const { name, description, waitInterval } = table;
      if (!tables[name]) {
        tables[name] = {
          name,
          description,
          waitInterval,
          sites: [siteCode],
        };
      } else if (!tables[name].sites.includes(siteCode)) {
        tables[name].sites.push(siteCode);
      }
    });
  });
  const digits = Math.floor(Math.log10(Object.keys(tables).length) + 1);
  Object.keys(tables).forEach((table, idx) => {
    tables[table].tableCode = `T${(idx + 1).toString().padStart(digits, '0')}`;
  });
  return tables;
};

/**
  Calculate Rows
  Take a state object and spit out a new state containing rows for use in the SVG grid
*/
const calculateRows = (state) => {
  if (!state.neonContextLoaded) { return state; }
  const newState = {
    ...state,
    rows: {},
    rowLabels: [],
    rowTitles: {},
  };
  const {
    sites,
    allSites,
    allStates,
    allDomains,
    breakouts,
    sortDirection,
    tables: knownTables,
  } = state;
  if (!breakouts.length) {
    newState.rowLabels = ['ALL'];
    newState.rows.ALL = {};
  }
  // Generate unsorted row labels and row data respecting breakouts
  sites.forEach((site) => {
    const { siteCode, tables } = site;
    if (!allSites[siteCode]) { return; }
    const { stateCode, domainCode } = allSites[siteCode];
    tables.forEach((table) => {
      const { name: tableName, months } = table;
      if (!knownTables[tableName]) { return; }
      const { tableCode } = knownTables[tableName];
      // turn siteCode / stateCode / domainCode / tableCode into a label
      // then populate that label's row
      let title = !breakouts.length ? 'Summary of all sites and tables' : '';
      const label = !breakouts.length ? 'ALL' : (
        breakouts.reduce((acc, cur) => {
          switch (cur) {
            case 'states':
              title = `${title}${title.length ? ' - ' : ''}${allStates[stateCode].name}`;
              return `${acc}${acc.length ? '-' : ''}${stateCode}`;
            case 'domains':
              title = `${title}${title.length ? ' - ' : ''}${allDomains[domainCode].name}`;
              return `${acc}${acc.length ? '-' : ''}${domainCode}`;
            case 'sites':
              title = `${title}${title.length ? ' - ' : ''}${allSites[siteCode].description}`;
              return `${acc}${acc.length ? '-' : ''}${siteCode}`;
            case 'tables':
              title = `${title}${title.length ? ' - ' : ''}${tableName}`;
              return `${acc}${acc.length ? '-' : ''}${tableCode}`;
            default:
              return acc;
          }
        }, '')
      );
      if (!newState.rowLabels.includes(label)) {
        newState.rowLabels.push(label);
        newState.rows[label] = {};
      }
      newState.rowTitles[label] = title;
      Object.keys(months).forEach((month) => {
        newState.rows[label][month] = [...(newState.rows[label][month] || []), months[month]];
      });
    });
    // Sort row labels
    const sortReturns = [
      sortDirection === SORT_DIRECTIONS.ASC ? 1 : -1,
      sortDirection === SORT_DIRECTIONS.ASC ? -1 : 1,
    ];
    newState.rowLabels.sort((a, b) => (a < b ? sortReturns[0] : sortReturns[1]));
  });
  return newState;
};

/**
   Reducer
*/
const reducer = (state, action) => {
  const newState = { ...state };
  switch (action.type) {
    case 'setNeonContextValues':
      newState.allSites = action.allSites || {};
      newState.allStates = action.allStates || {};
      newState.allDomains = action.allDomains || {};
      newState.neonContextLoaded = true;
      return calculateRows(newState);

    case 'setBreakouts':
      if (!action.breakouts.every(b => state.validBreakouts.includes(b))) { return state; }
      newState.breakouts = [...action.breakouts];
      // If new breakouts contains the old leading breakout (the sort method) then keep it in front
      if (state.breakouts.length && newState.breakouts.includes(state.breakouts[0])) {
        newState.breakouts = newState.breakouts.filter(b => b !== state.breakouts[0]);
        newState.breakouts.unshift(state.breakouts[0]);
      }
      return calculateRows(newState);

    case 'setSortMethod':
      // Sort method is not a top-level value; rather the first breakout in the list.
      if (!state.validBreakouts.includes(action.method)) { return state; }
      newState.breakouts = state.breakouts.filter(b => b !== action.method);
      newState.breakouts.unshift(action.method);
      return calculateRows(newState);

    case 'setSortDirection':
      if (!Object.keys(SORT_DIRECTIONS).includes(action.direction)) { return state; }
      newState.sortDirection = action.direction;
      return calculateRows(newState);

    // Default
    default:
      return state;
  }
};

/**
   Context and Hook
*/
const Context = createContext(DEFAULT_STATE);
const useAvailabilityState = () => {
  const hookResponse = useContext(Context);
  if (hookResponse.length !== 2) {
    return [cloneDeep(DEFAULT_STATE), () => {}];
  }
  return hookResponse;
};

/**
   Context Provider
*/
const Provider = (props) => {
  const { sites, children } = props;

  const [
    { data: neonContextData, hasLoaded: neonContextHasLoaded },
  ] = NeonContext.useNeonContextState();

  /**
     Initial State and Reducer Setup
  */
  const initialState = { ...cloneDeep(DEFAULT_STATE), sites };
  initialState.tables = extractTables(initialState);
  const [state, dispatch] = useReducer(reducer, calculateRows(initialState));

  /**
     Effect - Watch for changes to NeonContext data and push into local state
  */
  useEffect(() => {
    if (!(neonContextHasLoaded && !state.neonContextLoaded)) { return; }
    const { sites: allSites, states: allStates, domains: allDomains } = neonContextData;
    dispatch({
      type: 'setNeonContextValues',
      allSites,
      allStates,
      allDomains,
    });
  }, [
    neonContextData,
    neonContextHasLoaded,
    state.neonContextLoaded,
    dispatch,
  ]);

  /**
     Render
  */
  return (
    <Context.Provider value={[state, dispatch]}>
      {children}
    </Context.Provider>
  );
};

Provider.propTypes = {
  sites: AvailabilityPropTypes.enhancedSites,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.string,
    ])),
    PropTypes.node,
    PropTypes.string,
  ]).isRequired,
};

Provider.defaultProps = {
  sites: [],
};


/**
   Export
*/
const AvailabilityContext = {
  Provider,
  useAvailabilityState,
  SORT_DIRECTIONS,
};

export default AvailabilityContext;
