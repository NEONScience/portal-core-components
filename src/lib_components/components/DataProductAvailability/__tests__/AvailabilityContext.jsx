import React from 'react';
import renderer from 'react-test-renderer';
import { renderHook } from '@testing-library/react-hooks';

import cloneDeep from 'lodash/cloneDeep';

import sitesJSON from '../../../../sampleData/sites.json';
import statesJSON from '../../../staticJSON/states.json';
import domainsJSON from '../../../staticJSON/domains.json';

import AvailabilityContext, { getTestableItems } from '../AvailabilityContext';

const {
  Provider,
  useAvailabilityState,
  SORT_DIRECTIONS,
} = AvailabilityContext;

const {
  DEFAULT_STATE,
  calculateRows,
  extractTables,
  hydrateNeonContextData,
  reducer,
} = getTestableItems();

const initialHydratedState = {
  ...cloneDeep(DEFAULT_STATE),
  neonContextHydrated: true,
  reference: {
    sites: sitesJSON,
    states: statesJSON,
    domains: domainsJSON,
  },
};

describe('DataProductAvailability - AvailabilityContext', () => {
  describe('Provider', () => {
    test('renders with no props', (done) => {
      setTimeout(() => {
        const tree = renderer
          .create(
            <Provider>
              <div>children</div>
            </Provider>,
          ).toJSON();
        expect(tree).toMatchSnapshot();
        done();
      });
    });
  });

  describe('useAvailabilityState()', () => {
    test('returns default state and a passthough when invoked outside of a provider', () => {
      const { result } = renderHook(() => useAvailabilityState());
      expect(Array.isArray(result.current)).toBe(true);
      expect(result.current.length).toBe(2);
      const [state, dispatch] = result.current;
      expect(state).toStrictEqual(DEFAULT_STATE);
      expect(typeof dispatch).toBe('function');
      expect(dispatch()).toBeUndefined();
    });
  });

  describe('calculateRows()', () => {
    test('does nothing if NeonContext is not yet hydrated in state', () => {
      expect(calculateRows(DEFAULT_STATE)).toStrictEqual(DEFAULT_STATE);
    });
    const initialState = {
      ...cloneDeep(initialHydratedState),
      rows: { foo: 'bar' },
      rowTitles: { baz: 'qux' },
      rowLabels: ['foo', 'bar'],
      sites: [
        {
          siteCode: 'ABBY',
          tables: [ // WA / D16
            { name: 't1', months: { '2010-01': 'available', '2010-10': 'delayed' } },
            { name: 't2', months: { '2010-01': 'available', '2010-10': 'tentative' } },
          ],
        },
        {
          siteCode: 'BARR',
          tables: [ // AK / D18
            { name: 't1', months: { '2010-10': 'not collected', '2013-10': 'expected' } },
            { name: 't3', months: { '2010-10': 'delayed', '2013-10': 'available' } },
          ],
        },
        {
          siteCode: 'CLBJ',
          tables: [ // TX / D11
            { name: 't2', months: { '2013-10': 'delayed', '2017-10': 'expected' } },
            { name: 't3', months: { '2013-10': 'tentative', '2017-10': 'being processed' } },
            { name: 'unknown table' },
          ],
        },
        {
          siteCode: 'INVALID', tables: [],
        },
      ],
      tables: {
        t1: { name: 't1', sites: ['ABBY', 'BARR'], tableCode: 'T1' },
        t2: { name: 't2', sites: ['ABBY', 'CLBJ'], tableCode: 'T2' },
        t3: { name: 't3', sites: ['BARR', 'CLBJ'], tableCode: 'T3' },
      },
    };
    test('appropriately calculates rows, rowTitles, and rowLabels for no breakouts (ALL)', () => {
      const expectedOutput = {
        rows: {
          ALL: {
            '2010-01': ['available', 'available'],
            '2010-10': ['delayed', 'tentative', 'not collected', 'delayed'],
            '2013-10': ['expected', 'available', 'delayed', 'tentative'],
            '2017-10': ['expected', 'being processed'],
          },
        },
        rowTitles: {
          ALL: 'Summary of all sites and tables',
        },
        rowLabels: ['ALL'],
      };
      const newState = calculateRows({ ...initialState, breakouts: [] });
      expect(newState.rows).toStrictEqual(expectedOutput.rows);
      expect(newState.rowTitles).toStrictEqual(expectedOutput.rowTitles);
      expect(newState.rowLabels).toStrictEqual(expectedOutput.rowLabels);
    });
    test('appropriately calculates rows, rowTitles, and rowLabels for discrete breakouts', () => {
      const expectedOutput = {
        rows: {
          'AK-D18-BARR-T1': {
            '2010-10': ['not collected'],
            '2013-10': ['expected'],
          },
          'AK-D18-BARR-T3': {
            '2010-10': ['delayed'],
            '2013-10': ['available'],
          },
          'TX-D11-CLBJ-T2': {
            '2013-10': ['delayed'],
            '2017-10': ['expected'],
          },
          'TX-D11-CLBJ-T3': {
            '2013-10': ['tentative'],
            '2017-10': ['being processed'],
          },
          'WA-D16-ABBY-T1': {
            '2010-01': ['available'],
            '2010-10': ['delayed'],
          },
          'WA-D16-ABBY-T2': {
            '2010-01': ['available'],
            '2010-10': ['tentative'],
          },
        },
        rowTitles: {
          'AK-D18-BARR-T1': 'Alaska - Tundra - Utqiaġvik - t1',
          'AK-D18-BARR-T3': 'Alaska - Tundra - Utqiaġvik - t3',
          'TX-D11-CLBJ-T2': 'Texas - Southern Plains - LBJ National Grassland - t2',
          'TX-D11-CLBJ-T3': 'Texas - Southern Plains - LBJ National Grassland - t3',
          'WA-D16-ABBY-T1': 'Washington - Pacific Northwest - Abby Road - t1',
          'WA-D16-ABBY-T2': 'Washington - Pacific Northwest - Abby Road - t2',
        },
        rowLabels: [
          'WA-D16-ABBY-T2',
          'WA-D16-ABBY-T1',
          'TX-D11-CLBJ-T3',
          'TX-D11-CLBJ-T2',
          'AK-D18-BARR-T3',
          'AK-D18-BARR-T1',
        ],
      };
      const newState = calculateRows({
        ...initialState,
        breakouts: ['states', 'domains', 'sites', 'foo', 'tables'],
      });
      expect(newState.rows).toStrictEqual(expectedOutput.rows);
      expect(newState.rowTitles).toStrictEqual(expectedOutput.rowTitles);
      expect(newState.rowLabels).toStrictEqual(expectedOutput.rowLabels);
    });
  });

  describe('extractTables()', () => {
    const initialState = {
      ...cloneDeep(initialHydratedState),
      sites: [
        {
          siteCode: 'ABBY',
          tables: [
            { name: 't1', description: 'table 1', waitInterval: 60 },
            { name: 't2', description: 'table 2', waitInterval: 40 },
            { name: 't1', description: 'table 1', waitInterval: 60 },
          ],
        },
        {
          siteCode: 'BARR',
          tables: [
            { name: 't1', description: 'table 1 different description', waitInterval: 60 },
            { name: 't3', description: 'table 3', waitInterval: 100 },
          ],
        },
      ],
    };
    const expectedOutput = {
      t1: {
        name: 't1',
        description: 'table 1',
        waitInterval: 60,
        sites: ['ABBY', 'BARR'],
        tableCode: 'T1',
      },
      t2: {
        name: 't2',
        description: 'table 2',
        waitInterval: 40,
        sites: ['ABBY'],
        tableCode: 'T2',
      },
      t3: {
        name: 't3',
        description: 'table 3',
        waitInterval: 100,
        sites: ['BARR'],
        tableCode: 'T3',
      },
    };
    test('properly generates extracted tables object', () => {
      expect(extractTables(initialState)).toStrictEqual(expectedOutput);
    });
  });

  describe('hydrateNeonContextData()', () => {
    test('properly applies NeonContext data', () => {
      const newState = hydrateNeonContextData(
        cloneDeep(DEFAULT_STATE),
        {
          sites: { foo: 'bar' },
          states: { abc: 'def' },
          domains: { ghi: 'jkl' },
          otherContextData: 'foo',
        },
      );
      expect(newState.reference).toStrictEqual({
        sites: { foo: 'bar' },
        states: { abc: 'def' },
        domains: { ghi: 'jkl' },
      });
      expect(newState.neonContextHydrated).toBe(true);
      Object.keys(newState)
        .filter((key) => !['reference', 'neonContextHydrated'].includes(key))
        .forEach((key) => {
          expect(newState[key]).toStrictEqual(DEFAULT_STATE[key]);
        });
    });
  });

  describe('reducer()', () => {
    test('reflects back original state for unrecognized action', () => {
      expect(
        reducer({ foo: 'bar' }, { type: 'unknown' }),
      ).toStrictEqual({ foo: 'bar' });
    });
    describe('hydrateNeonContextData', () => {
      test('requires neonContextData in the action to do anything', () => {
        expect(
          reducer({ foo: 'bar' }, { type: 'hydrateNeonContextData' }),
        ).toStrictEqual({ foo: 'bar' });
      });
      test('hydrates data and calculates rows if neonContextData is present', () => {
        const initialState = cloneDeep(DEFAULT_STATE);
        const neonContextData = {
          sites: sitesJSON,
          states: statesJSON,
          domains: domainsJSON,
        };
        const newState = reducer(initialState, { type: 'hydrateNeonContextData', neonContextData });
        expect(newState.neonContextHydrated).toBe(true);
        expect(newState.reference).toStrictEqual(neonContextData);
        expect(newState.rows).toStrictEqual({ ALL: {} });
      });
    });
    describe('setBreakouts', () => {
      test('requires action breakouts to be an array of all valid breakouts', () => {
        expect(
          reducer(initialHydratedState, { type: 'setBreakouts' }),
        ).toStrictEqual(initialHydratedState);
        expect(
          reducer(initialHydratedState, { type: 'setBreakouts', breakouts: ['states', 'qux'] }),
        ).toStrictEqual(initialHydratedState);
      });
      test('sets breakouts and caluclates rows', () => {
        const newState = reducer(
          initialHydratedState,
          { type: 'setBreakouts', breakouts: ['states', 'domains'] },
        );
        expect(newState.breakouts).toStrictEqual(['states', 'domains']);
      });
      test('keeps leading breakout in the front if still present in action breakouts', () => {
        const previousState = {
          ...initialHydratedState,
          breakouts: ['states', 'domains'],
        };
        const newState = reducer(
          previousState,
          { type: 'setBreakouts', breakouts: ['sites', 'domains', 'states'] },
        );
        expect(newState.breakouts).toStrictEqual(['states', 'sites', 'domains']);
      });
    });
    describe('setSortMethod', () => {
      test('requires action method be a valid breakout', () => {
        expect(
          reducer(initialHydratedState, { type: 'setSortMethod' }),
        ).toStrictEqual(initialHydratedState);
        expect(
          reducer(initialHydratedState, { type: 'setSortMethod', method: 'qux' }),
        ).toStrictEqual(initialHydratedState);
      });
      test('adds valid action method to the front of the breakout list', () => {
        const previousState = {
          ...initialHydratedState,
          breakouts: ['states', 'domains', 'tables'],
        };
        const newState = reducer(
          previousState,
          { type: 'setSortMethod', method: 'tables' },
        );
        expect(newState.breakouts).toStrictEqual(['tables', 'states', 'domains']);
      });
    });
    describe('setSortDirection', () => {
      test('requires a valid sort direction', () => {
        expect(
          reducer(initialHydratedState, { type: 'setSortDirection' }),
        ).toStrictEqual(initialHydratedState);
        expect(
          reducer(initialHydratedState, { type: 'setSortDirection', direction: 'qux' }),
        ).toStrictEqual(initialHydratedState);
      });
      test('applies valid sort direction', () => {
        const previousState = {
          ...initialHydratedState,
          sortDirection: SORT_DIRECTIONS.ASC,
        };
        const newState = reducer(
          previousState,
          { type: 'setSortDirection', direction: SORT_DIRECTIONS.DESC },
        );
        expect(newState.sortDirection).toBe(SORT_DIRECTIONS.DESC);
      });
    });
  });
});
