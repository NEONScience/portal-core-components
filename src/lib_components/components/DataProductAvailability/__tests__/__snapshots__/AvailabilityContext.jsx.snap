// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataProductAvailability - AvailabilityContext Provider renders with no props 1`] = `
<div>
  children
</div>
`;
